### ISSUE TITLE
Multistaged Dockerfile build fails because of incorrect cache data

### ISSUE THREAD

**Issue by mildred** [containers/buildah#1134]
**Description**

When running `podman build` to build Dockerfiles with multistaged builds, the previous stages cache becomes invalid when later stages are modified before the `COPY --from` line. Then and `COPY --from` stops working because it cannot find the data. If the stage the `COPY --from` is copying from is updated as to invalidate cache, it works again.

**Steps to reproduce the issue:**

first, start a build:
```
$ date >a
$ cat >Dockerfile <<EOF
FROM scratch AS source
COPY a /a

FROM scratch
LABEL serial 1
COPY --from=source /a /a
EOF
$ podman build .
STEP 1: FROM scratch
STEP 2: COPY a /a
--> 0fbfa04572eabb18e3a3a0f779a20ec79dc156bc66050002f8da53082befd293
STEP 3: COMMIT 
STEP 4: FROM scratch
STEP 5: LABEL serial 1
--> Using cache faf4d5e2c159f8fa41b2e0cd34666f387e01bb9b06271d8e7381f90ed4fab70d
STEP 6: FROM faf4d5e2c159f8fa41b2e0cd34666f387e01bb9b06271d8e7381f90ed4fab70d
STEP 7: COPY --from=source /a /a
--> 126a3a7448fb2597335723510fb268d0953134ea423abe3cc4be2e9af774dda5
STEP 8: COMMIT 
```

Update the second stage before the `COPY --from`

```
$ cat >Dockerfile <<EOF
FROM scratch AS source
COPY a /a

FROM scratch
LABEL serial 2
COPY --from=source /a /a
EOF
$ podman build .
STEP 1: FROM scratch
STEP 2: COPY a /a
--> Using cache 0fbfa04572eabb18e3a3a0f779a20ec79dc156bc66050002f8da53082befd293
STEP 3: COMMIT 
STEP 4: FROM scratch
STEP 5: LABEL serial 2
--> Using cache 0a99b66dbe84a9423ad798495324e03255dfbb3ea157afe820e037a2077e42cf
STEP 6: FROM 0a99b66dbe84a9423ad798495324e03255dfbb3ea157afe820e037a2077e42cf
STEP 7: COPY --from=source /a /a
error building at step {Env:[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:copy Args:[/a /a] Flags:[--from=source] Attrs:map[] Message:COPY --from=source /a /a Original:COPY --from=source /a /a}: no files found matching "/var/home/mildred/.local/share/containers/storage/vfs/dir/313aed8d769c90f1c7075099565d058cae01b02f20f63946d4dc268e3810ef0a/a": no such file or directory
```

It is possible to workaround this by updating the first stage:

```
$ cat >Dockerfile <<EOF
FROM scratch AS source
COPY a /a
LABEL serial 1

FROM scratch
LABEL serial 2
COPY --from=source /a /a
EOF
$ podman build .
STEP 1: FROM scratch
STEP 2: COPY a /a
--> Using cache 0fbfa04572eabb18e3a3a0f779a20ec79dc156bc66050002f8da53082befd293
STEP 3: FROM 0fbfa04572eabb18e3a3a0f779a20ec79dc156bc66050002f8da53082befd293
STEP 4: LABEL serial 1
--> 2e18d677899b10e1218ce148483644c3b7b13d1a25b95f2f528e444f61de3c9b
STEP 5: COMMIT 
STEP 6: FROM scratch
STEP 7: LABEL serial 2
--> Using cache 0a99b66dbe84a9423ad798495324e03255dfbb3ea157afe820e037a2077e42cf
STEP 8: FROM 0a99b66dbe84a9423ad798495324e03255dfbb3ea157afe820e037a2077e42cf
STEP 9: COPY --from=source /a /a
--> 53ffcce9e7f4c64d98930e1efb456f121c41f19beb7076400e732f49152f8335
STEP 10: COMMIT 
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
podman-0.10.1-1.gite4a1553.fc29.x86_64
buildah-1.4-1.dev.git0a7389c.fc29.x86_64
```

**Output of `buildah version`:**

```
$ buildah version
strconv.ParseInt: parsing "": invalid syntax
$ buildah --version
buildah version 1.4-dev (image-spec 1.0.0, runtime-spec 1.0.0)
```

**Output of `podman version` if reporting a `podman build` issue:**

```
$ podman version
Version:       0.10.1
Go Version:    go1.11
OS/Arch:       linux/amd64
```

**Output of `cat /etc/*release`:**

```
Fedora release 29 (Twenty Nine)
NAME=Fedora
VERSION="29.1.2 (Workstation Edition)"
ID=fedora
VERSION_ID=29
PLATFORM_ID="platform:f29"
PRETTY_NAME="Fedora 29.1.2 (Workstation Edition)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:29"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f29/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=29
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=29
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
OSTREE_VERSION=29.1.2
Fedora release 29 (Twenty Nine)
Fedora release 29 (Twenty Nine)
```

**Output of `uname -a`:**

```
Linux elayna 4.18.16-300.fc29.x86_64 #1 SMP Sat Oct 20 23:24:08 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base.
# device.
# mkfsarg = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver 
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

# If specified, use OSTree to deduplicate files with the overlay backend
ostree_repo = ""

# Set to skip a PRIVATE bind mount on the storage home directory.  Only supported by
# certain container storage drivers
skip_mount_home = "false"
```

**Output of `cat ~/.config/containers/storage.conf`:**

```
[storage.options]
#ostree_repo = "/home/mildred/ostree"

```

**(added out of completeness, I suggest adding this to your issue template)**

**Comment by TomSweeneyRedHat** [comment]
@mildred Thanks for the problem report, we'll take a look.

**Comment by TomSweeneyRedHat** [comment]
@mildred we've made some changes recently in Buildah in how it handles FROM in Dockerfiles and master is now at:

```
# buildah version
Version:         1.5-dev
Go Version:      go1.9.7
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  v0.7.0-alpha1
Git Commit:      a34ca2ad
Built:           Sat Oct 27 17:11:03 2018
OS/Arch:         linux/amd64
```

The latest seems to work for me.  Can you verify that I tested in the right way and/or upgrade Buildah to the lastet and retry?  Test output follows.

```
# cat Dockerfile
FROM scratch AS source
COPY a /a

FROM scratch
LABEL serial 2 
COPY --from=source /a /a
[root@localhost ~]# buildah bud .
STEP 1: FROM scratch AS source
STEP 2: COPY a /a
STEP 3: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@efec61b6a284365342f0c1ec6de392afba023164cc023c4959cc095ddc63d2fe
Getting image source signatures
Skipping fetch of repeat blob sha256:adb4c62a40dc372b1bd680471049644484e91b8d82a9eb20f90593f9d3d7e39a
Copying config sha256:1be16fca8021a61cb8ca02b437177ef383b639b0f7d9ab3cb21f8cd8b185c8ed
 338 B / 338 B [============================================================] 0s
Writing manifest to image destination
Storing signatures
--> efec61b6a284365342f0c1ec6de392afba023164cc023c4959cc095ddc63d2fe
STEP 4: FROM scratch
STEP 5: LABEL serial 2 
STEP 6: COPY --from=source /a /a
STEP 7: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@56c275c3fc54a6cf24864e2835036b59b942aa7f640c74547bc2d60d5be192d8
Getting image source signatures
Skipping fetch of repeat blob sha256:adb4c62a40dc372b1bd680471049644484e91b8d82a9eb20f90593f9d3d7e39a
Copying config sha256:02bdfeecbe221dcc057b9930276b503a6d600eacb0c7b9c2b147d8f40ed8fba4
 362 B / 362 B [============================================================] 0s
Writing manifest to image destination
Storing signatures
--> 56c275c3fc54a6cf24864e2835036b59b942aa7f640c74547bc2d60d5be192d8
```

**Comment by rhatdan** [comment]
Fixed in buildah 1.5
